torch/fx/passes/runtime_assert.py

```diff
@@ -222,89 +306,171 @@ def match_symbol(symint, cb):
-                    last_sym_node = last_sym_node.next
-                with graph.inserting_before(last_sym_node.next):
-                    add_runtime_asserts(ras_by_symbol.pop(None, []))  # type: ignore[call-overload]
+            if node == first_non_placeholder:
+                add_runtime_asserts(ras_by_symbol.pop(None, []))  # type: ignore[call-overload]
+
+            # deduplicate asserts already present in graph
+            if node.target in (
+                torch._check,
+                torch.ops.aten._assert_scalar.default,
+            ):
+                if (
+                    node.args[0] == True  # trivial
+                    or (assert_expr := _get_sym_val(node.args[0])) in added_asserts
+                ):
+                    gm.graph.erase_node(node)
+                else:
+                    added_asserts.add(assert_expr)
+
+            # hash cons on symbolic values
+            if (
+                node.op == "call_function"
+                and (sym_expr := _get_sym_val(node)) is not None
+            ):
+                if (
+                    sym_expr in symbol_to_proxy  # example value is redundant
+                    or (
+                        _is_intermediate_tensor_shape_call(node)
+                        and not (sym_expr.free_symbols - symbol_to_proxy.keys())
+                    )  # shape call on intermediate tensor, turn into computation on input shapes
+                ):
+                    if _is_intermediate_tensor_shape_call(node):  # reify from input shapes
+                        symbol_to_proxy[sym_expr] = _sympy_interp(symbol_to_proxy, sym_expr)
+                        # won't try DCE-ing tensor compute here
+                    hash_node = symbol_to_proxy[sym_expr].node
+                    node.replace_all_uses_with(hash_node)
+                    gm.graph.erase_node(node)
+                    log.debug("CSE node %s -> %s for expr %s", node, hash_node, sym_expr)
+                else:
+                    symbol_to_proxy[sym_expr] = fx.Proxy(node)
+
+            if not any(ras for ras in ras_by_symbol.values()):
+                # if no runtime asserts are present, only do CSE,
+                # don't proceed further and add refinement calls
+                continue
+
+            # If the symbol is used, we'll call sym_constrain_range(_for_size) later when we see it anyways,
+            # so delete calls before that
+            if node.target in (
+                torch._check_is_size,
+                torch.ops.aten.sym_constrain_range.default,
+                torch.ops.aten.sym_constrain_range_for_size.default,
+            ):
+                gm.graph.erase_node(node)
 
             defs = []
 
             if unbacked_bindings := node.meta.get("unbacked_bindings"):
                 for s, keypath in unbacked_bindings.items():
                     defs.append(s)
-
-                    # TODO: some CSE when generating these nodes can probably
```

Comment:
Actually, it's not clear to me that you actually implemented the CSE that was specified here. We're talking about CSE when there is something like tensors[0].size(0) is u0, tensors[0].size(1), it could be profitable to CSE tensors[0]. I don't think this is a big deal but I think the comment should stay.

Reply:
Hmm, would you know of a good test case to try this out? I haven't really found export/dynamo test cases where the keypath length is > 1, so struggling to debug this

Reply:
I'd probably just leave the comment in here / clarify